### PR TITLE
Rider plugin fixes

### PR DESCRIPTION
**Pull request by mxprshn**
- Add test generation timeout option 
- Add balloon notifications
- Improve .NET versions check
- Pass multiple methods to V# API simultaneously
- Support types with overloaded methods
- Publish project before running VSharp

**Release notes:**
- Test generation timeout option to control the extensivity of test generation
- Balloon notifications informing about the test generation process state
- .NET 7 projects support
- Bugfixes

### CODE REVIEW

**Review by MchKosticyn — APPROVED** [review]

File: utbot-rider/src/dotnet/UtBot/UtBot/Utils/ProjectPublisher.cs [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+using JetBrains.Collections.Viewable;
+using JetBrains.Lifetimes;
+using JetBrains.ProjectModel;
+using JetBrains.ProjectModel.Properties;
+using JetBrains.Threading;
+using JetBrains.Util;
+
+namespace UtBot.Utils;
+
+[ShellComponent]
```
Comment (Domonion) [inline utbot-rider/src/dotnet/UtBot/UtBot/Utils/ProjectPublisher.cs:15]: why shell component?

File: utbot-rider/src/dotnet/UtBot/UtBot/Utils/ProjectPublisher.cs [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+public class ProjectPublisher
+{
+    private readonly LifetimeDefinition _lifetimeDef;
+
+    public ProjectPublisher(Lifetime lifetime)
+    {
+        _lifetimeDef = lifetime.CreateNested();
+    }
+
+    public void PublishSync(IProject project, IProjectConfiguration config, VirtualFileSystemPath outputDir)
```
Comment (Domonion) [inline utbot-rider/src/dotnet/UtBot/UtBot/Utils/ProjectPublisher.cs:25]: better make it `[ProjectComponent]` and receive project from constructor

File: utbot-rider/src/dotnet/UtBot/UtBot/Utils/Notifications.cs [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+    {
+        ShowNotification(
+            NotificationSeverity.WARNING,
+            Title,
+            body,
+            closeAfterExecution: closeAfterExecution,
+            command: command);
+    }
+
+    public void Refresh() => _lifetime = _sequentialLifetimes.Next();
```
Comment (Domonion) [inline utbot-rider/src/dotnet/UtBot/UtBot/Utils/Notifications.cs:99]: what is this for?

File: utbot-rider/src/dotnet/UtBot/UtBot/Utils/DotNetVersionUtils.cs [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+using System.Text.RegularExpressions;
+using JetBrains.ProjectModel;
+using JetBrains.Util;
+
+namespace UtBot.Utils;
+
+internal readonly record struct TestProjectTargetFramework(FrameworkMoniker FrameworkMoniker, bool IsDefault);
+
+[SolutionComponent]
+internal class DotNetVersionUtils
```
Comment (Domonion) [inline utbot-rider/src/dotnet/UtBot/UtBot/Utils/DotNetVersionUtils.cs:15]: Why internal class is SolutionComponent?

File: utbot-rider/src/dotnet/UtBot/UtBot/Utils/DotNetVersionUtils.cs [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+{
+    public const int MinCompatibleSdkMajor = 6;
+
+    private const string NUnitProjectMinTfm = "net6.0";
+
+    private readonly ConcurrentDictionary<string, TestProjectTargetFramework> _testProjectsFrameworksCache = new();
```
Comment (Domonion) [inline utbot-rider/src/dotnet/UtBot/UtBot/Utils/DotNetVersionUtils.cs:21]: to discuss - I believe we should not manually maintain cache as we do not track any refactorings and after some time of work this cache might grow indefinitely.

**Review by Domonion — CHANGES_REQUESTED** [review]
we need to discuss

File: utbot-rider/src/dotnet/UtBot/UtBot/UnitTestBuilder.cs [hunk 6]
```diff
@@ -91,66 +202,95 @@ protected override void Process(CSharpGeneratorContext context, IProgressIndicat
 
         project.Locks.AssertNonMainThread();
 
-        SolutionBuilderRequest buildRequest;
-        using (_shellLocks.UsingReadLock())
+        var config = project.ProjectProperties.ActiveConfigurations.Configurations.First();
+        var outputDir = project.GetOutputDirectory(config.TargetFrameworkId).Combine("utbot-publish");
+        progressIndicator.Header.SetValue($"Building {project.Name}...");
```
Comment (mxprshn) [inline utbot-rider/src/dotnet/UtBot/UtBot/UnitTestBuilder.cs:207]: Change method to indicate that 
1. you are doing publish
2. you are publishing for gathering dependencies
3. you are publishing to custom dir

Comment (mxprshn) [reply utbot-rider/src/dotnet/UtBot/UtBot/UnitTestBuilder.cs:207]: Tried this, rider can't show long strings properly in progress bars

Comment (Domonion) [reply utbot-rider/src/dotnet/UtBot/UtBot/UnitTestBuilder.cs]: make message shorter

File: utbot-rider/src/dotnet/UtBot/UtBot/Utils/ProjectPublisher.cs (hunk 1 above)
Comment (Domonion) [reply utbot-rider/src/dotnet/UtBot/UtBot/Utils/ProjectPublisher.cs]: remove all shell components please

File: utbot-rider/src/dotnet/UtBot/UtBot/UnitTestBuilder.cs (hunk 6 above)
Comment (mxprshn) [reply utbot-rider/src/dotnet/UtBot/UtBot/UnitTestBuilder.cs]: Already done. This screenshot is outdated

**Pull request merged**